kernel/src/syscall/sched_affinity.rs

```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::{cmp, mem};
+
+use ostd::cpu::{num_cpus, CpuId, CpuSet};
+
+use super::SyscallReturn;
+use crate::{prelude::*, process::posix_thread::thread_table, thread::Tid};
+
+pub fn sys_sched_getaffinity(
+    tid: Tid,
+    cpuset_size: usize,
+    cpu_set_ptr: Vaddr,
+    ctx: &Context,
+) -> Result<SyscallReturn> {
+    let cpu_set = match tid {
+        0 => ctx.thread.atomic_cpu_affinity().load(),
+        _ => match thread_table::get_thread(tid) {
+            Some(thread) => thread.atomic_cpu_affinity().load(),
+            None => return Err(Error::with_message(Errno::ESRCH, "thread does not exist")),
+        },
+    };
+
+    let bytes_written = write_cpu_set_to(ctx.get_user_space(), &cpu_set, cpuset_size, cpu_set_ptr)?;
+
+    Ok(SyscallReturn::Return(bytes_written as isize))
+}
+
+pub fn sys_sched_setaffinity(
+    tid: Tid,
+    cpuset_size: usize,
+    cpu_set_ptr: Vaddr,
+    ctx: &Context,
+) -> Result<SyscallReturn> {
+    let user_cpu_set = read_cpu_set_from(ctx.get_user_space(), cpuset_size, cpu_set_ptr)?;
+
+    match tid {
+        0 => ctx.thread.atomic_cpu_affinity().store(&user_cpu_set),
+        _ => match thread_table::get_thread(tid) {
+            Some(thread) => {
+                thread.atomic_cpu_affinity().store(&user_cpu_set);
+            }
+            None => return Err(Error::with_message(Errno::ESRCH, "thread does not exist")),
+        },
+    }
+
+    Ok(SyscallReturn::Return(0))
+}
+
+// Linux uses `DECLARE_BITMAP` for `cpu_set_t`, inside which each part is a
+// `long`. We use the same scheme to ensure byte endianness compatibility.
+type Part = u64;
+const SIZE_OF_PART: usize = mem::size_of::<Part>();
+const CPUS_IN_PART: usize = SIZE_OF_PART * 8;
+
```

Comment:
The `Part` and the subsequent read/write operations are somewhat difficult to understand. 

Would it be possible to define a struct `CCpuSet` that represents `cpu_set_t` for copying to and from user space? We would only need to implement the conversion between `CCpuSet` and `CpuSet`.

Reply:
Yes it is possible. But given that `cpu_set_t` is an intimidatingly long fix-sized 1024 bit array, I suppose that the current implementation is the most efficient approach and supports more than 1024 CPUs.